### PR TITLE
[Feature] add agency leader discovery

### PR DESCRIPTION
**Pull request by informalict**

### CODE REVIEW

File: pkg/deployment/agency/cache.go [hunk 1]
```diff
@@ -66,48 +69,135 @@ func (c cacheSingle) Data() (State, bool) {
...
+func (c *cache) GetLeaderID() string {
+	return c.leaderID
+}
+
+func (c *cache) Reload(ctx context.Context, clients []agency.Agency) (uint64, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	cfg, err := getAgencyConfig(ctx, client)
+	leaderCli, leaderConfig, err := getLeader(ctx, clients)
```
Comment (ajanikow) [inline pkg/deployment/agency/cache.go:99]: If one of Agents is unreachable - then ctx will be already expired. Provide a lot smaller ctx for config discovery

File: pkg/util/strings.go [hunk 2]
```diff
@@ -86,3 +86,14 @@ func DiffStrings(compareWhat, compareTo []string) []string {
 
 	return append(diff, DiffStringsOneWay(compareTo, compareWhat)...)
 }
+
+// IsStringInSlice returns true if haystack contains needle.
```
Comment (ajanikow) [inline pkg/util/strings.go:90]: USe instead StringListContains

File: pkg/deployment/agency/cache.go [hunk 3]
```diff
@@ -66,48 +69,135 @@ func (c cacheSingle) Data() (State, bool) {
...
-		c.commitIndex = cfg.CommitIndex
-		return cfg.CommitIndex, nil
+		c.commitIndex = leaderConfig.CommitIndex
+		return leaderConfig.CommitIndex, nil
+	}
+}
+
+// getLeader returns config and client to a leader agency.
+// If there is no quorum for the leader then error is returned.
+func getLeader(ctx context.Context, clients []agency.Agency) (agency.Agency, *agencyConfig, error) {
```
Comment (ajanikow) [inline pkg/deployment/agency/cache.go:129]: Add additional state if all agents are healthy & all have same commitIndex (so health)

**Review by ajanikow — CHANGES_REQUESTED** [review]

File: pkg/deployment/agency/cache.go (hunk 1 above)
Comment (informalict) [reply pkg/deployment/agency/cache.go]: done

File: pkg/util/strings.go (hunk 2 above)
Comment (informalict) [reply pkg/util/strings.go]: done

File: pkg/deployment/agency/cache.go (hunk 3 above)
Comment (informalict) [reply pkg/deployment/agency/cache.go]: done

**Review by ajanikow — APPROVED** [review]

**Pull request merged**